lib/url-decorator.jsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { CompositeDecorator } from 'draft-js';
+import urlRegex from 'url-regex';
+import normalizeUrl from 'normalize-url';
+import { viewExternalUrl } from './utils/url-utils';
+
+const urlRe = urlRegex();
+/*
+- markdownLinkRe is a regexp to be applied to the substring preceding a matched url, ie. `wholeString.slice(0, wholeString.indexOf(matchedUrl))`
+- it tests whether the matched url is preceded by the charcters that define a link in markdown, ie. `(` or `]: `
+*/
+const markdownLinkRe = /(?:\(|\]:\s+)$/;
+
+export function normalizeUrlsInMarkdownString( str ) {
+	return str.replace( urlRe, ( match, offset ) => {
+		var normalized = normalizeUrl( match );
+		if ( markdownLinkRe.test( str.slice( 0, offset ) ) ) {
+			return normalized;
+		}
+		// replace url with markdown link only if it's not already part of one
+		return `[${match}](${normalized})`;
+	} );
+}
+
+export const urlDecorator = new CompositeDecorator( [
+	{
+		strategy: function( contentBlock, callback ) {
+			const text = contentBlock.getText();
+			let matchArr, start;
+			while ( ( matchArr = urlRe.exec( text ) ) !== null ) {
+				start = matchArr.index;
+				callback( start, start + matchArr[0].length );
+			}
+		},
+
+		component: React.createClass( {
+			onClick: function( event ) {
+				event.preventDefault();
+				viewExternalUrl( normalizeUrl( this.props.decoratedText ) );
+			},
+
+			render: function() {
+				return (
+					<a
+						onClick={this.onClick}
+						href={this.props.decoratedText}>
+						{this.props.children}
+					</a>
+				);
+			}
+		} ),
```

Comment:
> could you expand on why you'd like this component to be defined elsewhere?

mainly I would like to see their definitions outside of the object passed in to create the decorator. if they are separate, we can test them (even though we don't have tests yet) and it can be a little easier to find them.

the call to create the decorator isn't that big of a deal either. if we think that something like this is confusion, it's probably more of a naming thing than anything else…

```js
new CompositeDecorator( {
	strategy: findLinks,
	component: Link
} );
```